OpenRA.Mods.Common/Traits/Production.cs

```diff
@@ -113,25 +113,63 @@ public virtual void DoProduction(Actor self, ActorInfo producee, ExitInfo exitin
 
 				var notifyOthers = self.World.ActorsWithTrait<INotifyOtherProduction>();
 				foreach (var notify in notifyOthers)
-					notify.Trait.UnitProducedByOther(notify.Actor, self, newUnit);
+					notify.Trait.UnitProducedByOther(notify.Actor, self, newUnit, productionType);
 
 				foreach (var t in newUnit.TraitsImplementing<INotifyBuildComplete>())
 					t.BuildingComplete(newUnit);
 			});
 		}
 
-		public virtual bool Produce(Actor self, ActorInfo producee, TypeDictionary inits)
+		protected virtual ExitInfo SelectExit(Actor self, ActorInfo producee, string productionType, Func<ExitInfo, bool> p)
+		{
+			if (productionType != null)
+			{
+				var exit = self.RandomExitOrDefault(productionType, p);
+				if (exit != null)
+					return exit;
+			}
+
+			var bi = producee.TraitInfoOrDefault<BuildableInfo>();
+			if (bi != null)
+			{
+				if (Info.Produces.Contains(bi.BuildAtProductionType))
+				{
+					var exit = self.RandomExitOrDefault(bi.BuildAtProductionType, p);
+					if (exit != null)
+						return exit;
+				}
+
+				foreach (var produce in Info.Produces)
+				{
+					if (bi.Queue.Contains(produce))
+					{
+						var exit = self.RandomExitOrDefault(produce, p);
+						if (exit != null)
+							return exit;
+					}
+				}
+			}
+
+			return self.RandomExitOrDefault(null, p);
```

Comment:
This unfortunately regresses behaviour when all exits are blocked.
The expected behaviour is for production to halt until there is space for the unit, which relies on this method returning `null`.  Returning a random blocked exit makes units stack on top of eachother.

Reply:
That has been there since I took over the PR. If you want it changed tell me how exactly it should behave.

Reply:
I'll need to invest some time to work out the intended purpose of the rest of this method, then will hopefully let you know.

Reply:
Ok, I think the following should do it:
```diff
diff --git a/OpenRA.Mods.Common/Traits/Buildings/Exit.cs b/OpenRA.Mods.Common/Traits/Buildings/Exit.cs
index a8f9abf54a..a97959f08c 100644
--- a/OpenRA.Mods.Common/Traits/Buildings/Exit.cs
+++ b/OpenRA.Mods.Common/Traits/Buildings/Exit.cs
@@ -40,22 +40,22 @@ public class Exit { }
 
        public static class ExitExts
        {
-               public static int CountExits(this ActorInfo info, string productionType = null)
-               {
-                       var all = info.TraitInfos<ExitInfo>();
-                       return productionType == null ? all.Count(e => e.ProductionTypes.Count == 0) : all.Count(e => e.ProductionTypes.Contains(productionType));
-               }
-
                public static ExitInfo FirstExitOrDefault(this ActorInfo info, string productionType = null)
                {
                        var all = info.TraitInfos<ExitInfo>();
-                       return productionType == null ? all.FirstOrDefault(e => e.ProductionTypes.Count == 0) : all.FirstOrDefault(e => e.ProductionTypes.Contains(productionType));
+                       if (productionType == null)
+                               return all.FirstOrDefault(e => e.ProductionTypes.Count == 0);
+
+                       return all.FirstOrDefault(e => e.ProductionTypes.Count == 0 || e.ProductionTypes.Contains(productionType));
                }
 
                public static IEnumerable<ExitInfo> Exits(this ActorInfo info, string productionType = null)
                {
                        var all = info.TraitInfos<ExitInfo>();
-                       return productionType == null ? all.Where(e => e.ProductionTypes.Count == 0) : all.Where(e => e.ProductionTypes.Contains(productionType));
+                       if (productionType == null)
+                               return all.Where(e => e.ProductionTypes.Count == 0);
+
+                       return all.Where(e => e.ProductionTypes.Count == 0 || e.ProductionTypes.Contains(productionType));
                }
 
                public static ExitInfo RandomExitOrDefault(this ActorInfo info, World world, string productionType, Func<ExitInfo, bool> p = null)
diff --git a/OpenRA.Mods.Common/Traits/Production.cs b/OpenRA.Mods.Common/Traits/Production.cs
index 1b0f80d3b2..d924de9743 100644
--- a/OpenRA.Mods.Common/Traits/Production.cs
+++ b/OpenRA.Mods.Common/Traits/Production.cs
@@ -122,35 +122,7 @@ public virtual void DoProduction(Actor self, ActorInfo producee, ExitInfo exitin
 
                protected virtual ExitInfo SelectExit(Actor self, ActorInfo producee, string productionType, Func<ExitInfo, bool> p)
                {
-                       if (productionType != null)
-                       {
-                               var exit = self.RandomExitOrDefault(productionType, p);
-                               if (exit != null)
-                                       return exit;
-                       }
-
-                       var bi = producee.TraitInfoOrDefault<BuildableInfo>();
-                       if (bi != null)
-                       {
-                               if (Info.Produces.Contains(bi.BuildAtProductionType))
-                               {
-                                       var exit = self.RandomExitOrDefault(bi.BuildAtProductionType, p);
-                                       if (exit != null)
-                                               return exit;
-                               }
-
-                               foreach (var produce in Info.Produces)
-                               {
-                                       if (bi.Queue.Contains(produce))
-                                       {
-                                               var exit = self.RandomExitOrDefault(produce, p);
-                                               if (exit != null)
-                                                       return exit;
-                                       }
-                               }
-                       }
-
-                       return self.RandomExitOrDefault(null, p);
+                       return self.RandomExitOrDefault(productionType, p);
                }
 
                protected ExitInfo SelectExit(Actor self, ActorInfo producee, string productionType)
```

But it would be good to get a second opinion from another reviewer before merging.

Reply:
Ok, once you are sure what it should look like I will implement it.

Reply:
I think it would be best to implement it now, and then it can be verified during the second review.